src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs

```diff
@@ -270,44 +271,99 @@ public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string full
-            if (searchPattern.Equals("."))
-                fileName = "*"; // A period is equivalent to an asterix to indicate a wildcard search.
-            string aqs = string.Format("System.FileName:~\"{0}\"", fileName);
-            query.ApplicationSearchFilter = aqs;
+            // Construct an AQS filter
+            var aqsRes = ConstructAQSQuery(path, searchPattern);
+            query.ApplicationSearchFilter = aqsRes.Item1;
+            string actualPath = aqsRes.Item2;
+
+            // If the filtered path is deeper than the given user path, we need to get a new folder for it.
+            // This occurs when someone does something like Enumerate("C:\first\second\", "C:\first\second\third\*").
+            // When AllDirectories is set this isn't an issue, but for TopDirectoryOnly we have to do some special work
+            // to make sure something is actually returned when the searchPattern is a subdirectory of the path.
+            // To do this, we attempt to get a new StorageFolder for the subdirectory and return an empty enumerable
+            // if we can't.
+            if (actualPath != folder.Path)
+            {
+                try
+                {
+                    folder = await StorageFolder.GetFolderFromPathAsync(Path.GetFullPath(actualPath)).TranslateWinRTTask(actualPath, isDirectory: true);
+                }
+                catch (FileNotFoundException) { return new IStorageItem[0]; }
+            }
 
-            // Execute CreateFileQueryWithOptions with our built query
-            List<IStorageItem> items = new List<IStorageItem>();
-            if ((searchTarget & SearchTarget.Files) > 0)
+            // Execute our built query
+            if (searchTarget == SearchTarget.Files)
             {
                 StorageFileQueryResult queryResult = folder.CreateFileQueryWithOptions(query);
-                items.AddRange(await queryResult.GetFilesAsync());
+                return await queryResult.GetFilesAsync().TranslateWinRTTask(folder.Path, isDirectory: true);
             }
-            // Execute CreateFolderQueryWithOptions with our built query
-            if ((searchTarget & SearchTarget.Directories) > 0)
+            else if (searchTarget == SearchTarget.Directories)
             {
                 StorageFolderQueryResult queryResult = folder.CreateFolderQueryWithOptions(query);
-                items.AddRange(await queryResult.GetFoldersAsync());
+                return await queryResult.GetFoldersAsync().TranslateWinRTTask(folder.Path, isDirectory: true);
+            }
+            else
+            {
+                StorageItemQueryResult queryResult = folder.CreateItemQueryWithOptions(query);
+                return await queryResult.GetItemsAsync().TranslateWinRTTask(folder.Path, isDirectory: true);
+            }
+        }
+
+        private static Tuple<string, string> ConstructAQSQuery(string path, string searchPattern)
```

Comment:
>Tuple is another allocation and rather unreadable, can you just use out here instead or break this into two functions if you can come up with logical separation.

I can break it into two functions.

>Also, you can avoid two calls to GetFolderFromPathAsync by reordering some of this code?

I thought about that but decided to leave it in to keep argument validation the same as Win32.

Example:
```EnumerateFileQuery("c:\temp\doesntexist", "c:\temp\doesntexist\reallydoesntexist\*")```
If we replace the first folder query with the second sub-query in this case, our behavior changes from throwing a DirectoryNotFoundException to returning an empty enumerable.

We could do Directory.Exists on the original path I suppose, but I don't know if that's any better than GetFolderFromPathAsync. I assumed it wouldn't be.